cranelift-codegen/src/isa/x86/fde.rs

```diff
@@ -0,0 +1,354 @@
+            CallFrameInstruction::Offset(X86_64::RA, cfa_offset)
+        }
+        _ => {
+            return None;
+        }
+    })
+}
+
+/// Creates FDE structure from FrameLayout.
+pub fn emit_fde(func: &Function, isa: &dyn TargetIsa, sink: &mut dyn FrameUnwindSink) {
+    assert!(isa.name() == "x86");
+
+    // Expecting function with System V prologue
+    assert!(
+        func.signature.call_conv == CallConv::Fast
+            || func.signature.call_conv == CallConv::Cold
+            || func.signature.call_conv == CallConv::SystemV
+    );
+
+    assert!(func.frame_layout.is_some(), "expected func.frame_layout");
+    let frame_layout = func.frame_layout.as_ref().unwrap();
+
+    let mut ebbs = func.layout.ebbs().collect::<Vec<_>>();
+    ebbs.sort_by_key(|ebb| func.offsets[*ebb]); // Ensure inst offsets always increase
+
+    let encinfo = isa.encoding_info();
+    let mut last_offset = 0;
+    let mut changes = Vec::new();
+    for ebb in ebbs {
+        for (offset, inst, size) in func.inst_offsets(ebb, &encinfo) {
+            let address_offset = (offset + size) as usize;
+            assert!(last_offset <= address_offset);
+            if let Some(cmds) = frame_layout.instructions.get(&inst) {
+                for cmd in cmds.iter() {
+                    changes.push((address_offset, cmd.clone()));
+                }
+            }
+            last_offset = address_offset;
+        }
+    }
+
+    let len = last_offset as u32;
+
+    let word_size = isa.pointer_bytes() as i32;
+
+    let encoding = Encoding {
+        format: Format::Dwarf32,
+        version: 1,
+        address_size: word_size as u8,
+    };
+    let mut frames = FrameTable::default();
+
+    let mut cfa_def_reg = return_address_reg(isa);
+    let mut cfa_def_offset = 0i32;
+
+    let mut cie = CommonInformationEntry::new(
+        encoding,
+        /* code_alignment_factor = */ 1,
+        /* data_alignment_factor = */ -word_size as i8,
+        return_address_reg(isa),
+    );
+    for ch in frame_layout.initial.iter() {
+        if let Some(cfi) = to_cfi(isa, ch, &mut cfa_def_reg, &mut cfa_def_offset) {
+            cie.add_instruction(cfi);
+        }
+    }
+
+    let cie_id = frames.add_cie(cie);
+
+    let mut fde = FrameDescriptionEntry::new(FUNCTION_ENTRY_ADDRESS, len);
+
+    for (addr, ch) in changes.iter() {
+        if let Some(cfi) = to_cfi(isa, ch, &mut cfa_def_reg, &mut cfa_def_offset) {
+            fde.add_instruction((*addr) as u32, cfi);
+        }
+    }
+
+    frames.add_fde(cie_id, fde);
+
+    let mut eh_frame = EhFrame::from(FDEWriter::new());
+    frames.write_eh_frame(&mut eh_frame).unwrap();
+
+    let (bytes, relocs) = eh_frame.clone().into_vec_and_relocs();
+
+    let unwind_start = sink.len();
+    sink.bytes(&bytes);
+
+    for (off, r) in relocs {
+        sink.reloc(r, off + unwind_start);
+    }
+    let fde_offset = unsafe { ptr::read::<u32>(bytes.as_ptr() as *const u32) } as usize + 4;
```

Comment:
I'm not sure if this just snuck by in review or not, but I'd at the very least want an explanation next to this pointer read with why it will be safe. I'd like to replace this with `u32::from_le_bytes(bytes.as_slice()[..8].try_into())` - since the checks here aren't in a particularly hot path it seems like that shouldn't be an issue?

Reply:
This whole thing is fighting against the gimli API a bit. You're using `FrameTable` to write a single CIE and FDE, and then trying to figure out the FDE offset.

How about if gimli made the CIE/FDE write methods public, and then you can call those methods instead of using `FrameTable`, and get the FDE offset simply as `bytes.len()` prior to calling the FDE write?